pkg/egressgateway/policy.go

```diff
@@ -177,33 +179,114 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
+		dstCIDRMinusExcludedCIDRs := []*net.IPNet{dstCIDR}
+		for _, excludedCIDR := range config.excludedCIDRs {
+			newDstCIDRMinuxExcludedCIDRs := []*net.IPNet{}
+			for _, cidr := range dstCIDRMinusExcludedCIDRs {
+				r, _, l := ip.PartitionCIDR(*cidr, *excludedCIDR)
+				newDstCIDRMinuxExcludedCIDRs = append(newDstCIDRMinuxExcludedCIDRs, append(r, l...)...)
+			}
+
+			dstCIDRMinusExcludedCIDRs = newDstCIDRMinuxExcludedCIDRs
+		}
+
+		cidrs = append(cidrs, dstCIDRMinusExcludedCIDRs...)
+	}
+
+	return cidrs
+}
+
 func (config *PolicyConfig) forEachEndpointAndDestination(epDataStore map[endpointID]*endpointMetadata,
-	f func(net.IP, *net.IPNet, *gatewayConfig)) {
+	f func(net.IP, *net.IPNet, bool, *gatewayConfig)) {
 
 	for _, endpoint := range epDataStore {
 		if !config.selectsEndpoint(endpoint) {
 			continue
 		}
 
 		for _, endpointIP := range endpoint.ips {
+			isExcludedCIDR := false
 			for _, dstCIDR := range config.dstCIDRs {
-				f(endpointIP, dstCIDR, &config.gatewayConfig)
+				f(endpointIP, dstCIDR, isExcludedCIDR, &config.gatewayConfig)
+			}
+
+			isExcludedCIDR = true
+			for _, excludedCIDR := range config.excludedCIDRs {
+				f(endpointIP, excludedCIDR, isExcludedCIDR, &config.gatewayConfig)
+			}
+		}
+	}
+}
+
+func (config *PolicyConfig) forEachEndpointAndDestinationMinusExcludedCIDRs(epDataStore map[endpointID]*endpointMetadata,
+	f func(net.IP, *net.IPNet, *gatewayConfig)) {
+
+	cidrs := config.destinationMinusExcludedCIDRs()
+
+	for _, endpoint := range epDataStore {
+		if !config.selectsEndpoint(endpoint) {
+			continue
+		}
+
+		for _, endpointIP := range endpoint.ips {
+			for _, cidr := range cidrs {
+				f(endpointIP, cidr, &config.gatewayConfig)
 			}
 		}
 	}
 }
 
 func (config *PolicyConfig) matches(epDataStore map[endpointID]*endpointMetadata,
-	f func(net.IP, *net.IPNet, *gatewayConfig) bool) bool {
+	f func(net.IP, *net.IPNet, bool, *gatewayConfig) bool) bool {
 
 	for _, endpoint := range epDataStore {
 		if !config.selectsEndpoint(endpoint) {
 			continue
 		}
 
 		for _, endpointIP := range endpoint.ips {
+			isExcludedCIDR := false
 			for _, dstCIDR := range config.dstCIDRs {
-				if f(endpointIP, dstCIDR, &config.gatewayConfig) {
+				if f(endpointIP, dstCIDR, isExcludedCIDR, &config.gatewayConfig) {
+					return true
+				}
+			}
+
+			isExcludedCIDR = true
+			for _, excludedCIDR := range config.excludedCIDRs {
+				if f(endpointIP, excludedCIDR, isExcludedCIDR, &config.gatewayConfig) {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+func (config *PolicyConfig) matchesMinusExcludedCIDRs(epDataStore map[endpointID]*endpointMetadata,
+	f func(net.IP, *net.IPNet, *gatewayConfig) bool) bool {
+
+	cidr := config.destinationMinusExcludedCIDRs()
+
+	for _, endpoint := range epDataStore {
+		if !config.selectsEndpoint(endpoint) {
+			continue
+		}
+
+		for _, endpointIP := range endpoint.ips {
+			for _, cidr := range cidr {
+				if f(endpointIP, cidr, &config.gatewayConfig) {
```

Comment:
I'm amazed by the `cidr := range cidr`. Can we give these two variables different names?

Reply:
I blame `:s/<something>/<something>/g` (the returned slice from `config.destinationMinusExcludedCIDRs()` should really be `cidrs`)